### PR TITLE
Make SLA performance counter accurate

### PR DESCRIPTION
**Pull request by andreasohlund**
By including the time it takes to batch the outgoing messages into account.

WIP: How do we deal with the 2 todos? Throw? (previous versions silently made the feature a no-op)

Documentation not needed since this makes the counters as accurate as in v5

Closes https://github.com/Particular/NServiceBus/issues/3565

### CODE REVIEW

**Comment by SimonCropp** [comment]
@andreasohlund re the two TODOs yeah i would throw

**Comment by andreasohlund** [comment]
Removed the todo's and throwing instead, WIP removed.

@Particular/nservicebus-maintainers please review

File: src/NServiceBus.Core/Performance/Statistics/SLA/SLAMonitoring.cs [hunk 1]
```diff
@@ -18,31 +23,180 @@ protected internal override void Setup(FeatureConfigurationContext context)
...
+                throw new Exception("SLA Monitoring is not supported for send only endpoints, please remove .EnableSLAPerformanceCounter(mySLA) from your config.");
             }
 
             TimeSpan endpointSla;
-            if (!TryGetSLA(context, out endpointSla))
+
+            if (!context.Settings.TryGet(EndpointSLAKey, out endpointSla))
             {
-                return;
+                throw new Exception("Endpoint SLA is required for the `SLA violation countdown` counter. Please pass the SLA for this endpoint to .EnableSLAPerformanceCounter(mySLA).");
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Performance/Statistics/SLA/SLAMonitoring.cs]: remove please :)

Comment (andreasohlund) [inline src/NServiceBus.Core/Performance/Statistics/SLA/SLAMonitoring.cs]: fixed

**Comment by timbussmann** [comment]
:+1:

**Comment by danielmarbach** [comment]
@andreasohlund I haven't fully thought it through since I was in a hurry

https://github.com/Particular/NServiceBus/pull/3754

What do you think of this?

* `DataPoint` can be a struct since we are only saving `TimeSpan` and `DateTime` which are also structs to a list which saves us allocations
* Minimizes the number of enumerations on the collection
* Avoid unnecessary boxing by not using `IList`

**Comment by danielmarbach** [comment]
@andreasohlund @Scooletz and I will send in another PR on top of this one soonish

**Comment by timbussmann** [comment]
should we tag this one as WIP meanwhile?

**Comment by danielmarbach** [comment]
@timbussmann the PR is already here #3762

**Comment by timbussmann** [comment]
as discussed with @andreasohlund and @Scooletz I'm merging this since https://github.com/Particular/NServiceBus/pull/3762 has been marked WIP again and will be re targeted by @Scooletz if necessary.

**Comment by danielmarbach** [comment]
Apparently I was just too late. I think we should revisit it and add more proper tests to verify the current behavior so that we can apply #3762

**Comment by timbussmann** [comment]
@danielmarbach is that something we can address as part of #3762?

**Pull request merged**